lightning/src/ln/payment_tests.rs

```diff
@@ -563,6 +563,228 @@ fn retry_with_no_persist() {
+	mine_transaction(&nodes[1], &bs_commitment_tx[0]);
+	if !use_dust {
+		connect_blocks(&nodes[0], TEST_FINAL_CLTV - 1 + (MIN_CLTV_EXPIRY_DELTA as u32));
+		connect_blocks(&nodes[1], TEST_FINAL_CLTV - 1 + (MIN_CLTV_EXPIRY_DELTA as u32));
+		let as_htlc_timeout = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().split_off(0);
+		check_spends!(as_htlc_timeout[0], bs_commitment_tx[0]);
+		assert_eq!(as_htlc_timeout.len(), 1);
+
+		mine_transaction(&nodes[0], &as_htlc_timeout[0]);
+		// nodes[0] may rebroadcast (or RBF-bump) its HTLC-Timeout, so wipe the announced set.
+		nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap().clear();
+		mine_transaction(&nodes[1], &as_htlc_timeout[0]);
+	}
+
+	// Create a new channel on which to retry the payment before we fail the payment via the
+	// HTLC-Timeout transaction. This avoids ChannelManager timing out the payment due to us
+	// connecting several blocks while creating the channel (implying time has passed).
+	// We do this with a zero-conf channel to avoid connecting blocks as a side-effect.
+	let (_, chan_id_3) = open_zero_conf_channel(&nodes[0], &nodes[1], None);
+	assert_eq!(nodes[0].node.list_usable_channels().len(), 1);
+
+	// If we attempt to retry prior to the HTLC-Timeout (or commitment transaction, for dust HTLCs)
+	// confirming, we will fail as its considered still-pending...
+	let (new_route, _, _, _) = get_route_and_payment_hash!(nodes[0], nodes[2], if use_dust { 1_000 } else { 1_000_000 });
+	assert!(nodes[0].node.retry_payment(&new_route, payment_id).is_err());
+	assert!(nodes[0].node.get_and_clear_pending_msg_events().is_empty());
+
+	// After ANTI_REORG_DELAY confirmations, the HTLC should be failed and we can try the payment
+	// again. We serialize the node first as we'll then test retrying the HTLC after a restart
+	// (which should also still work).
+	connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1);
+	connect_blocks(&nodes[1], ANTI_REORG_DELAY - 1);
+	// We set mpp_parts_remain to avoid having abandon_payment called
+	expect_payment_failed_conditions(&nodes[0], payment_hash, false, PaymentFailedConditions::new().mpp_parts_remain());
+
+	chan_0_monitor_serialized = test_utils::TestVecWriter(Vec::new());
+	get_monitor!(nodes[0], chan_id).write(&mut chan_0_monitor_serialized).unwrap();
+	chan_1_monitor_serialized = test_utils::TestVecWriter(Vec::new());
+	get_monitor!(nodes[0], chan_id_3).write(&mut chan_1_monitor_serialized).unwrap();
+	nodes_0_serialized = nodes[0].node.encode();
+
+	assert!(nodes[0].node.retry_payment(&new_route, payment_id).is_ok());
+	assert!(!nodes[0].node.get_and_clear_pending_msg_events().is_empty());
+
+	reload_node!(second_new_chain_monitor, second_nodes_0_deserialized, second_persister);
+	reconnect_nodes(&nodes[0], &nodes[1], (true, true), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (false, false));
+
+	// Now resend the payment, delivering the HTLC through to nodes[1] XXX
+	assert!(nodes[0].node.retry_payment(&new_route, payment_id).is_ok());
```

Comment:
Could add a comment that we previously would've failed here because the ChannelManager still would've seen the HTLC as pending (IIUC)?